tensorflow/python/autograph/operators/py_builtins.py

```diff
@@ -278,6 +281,24 @@ def raise_zero_rank_error():
                                raise_zero_rank_error)
 
 
+def _tf_dataset_len(s):
+  l = cardinality.cardinality(s)
+
+  def raise_infinite_cardinality_error():
+    msg = "len requires non-infinite dataset"
+    with ops.control_dependencies([control_flow_ops.Assert(False, [msg])]):
+      return constant_op.constant(-1, dtype=dtypes.int32)
+
+  def reduce_unknown_cardinality_dataset(d):
+    return d.reduce(constant_op.constant(0, dtype=dtypes.int32), lambda x, _: x + 1)
+
+  l = control_flow_ops.cond(l != cardinality.INFINITE, lambda: l, raise_infinite_cardinality_error)
+
```

Comment:
We should be able to make this even more efficient with something like this:

```
with ops.control_dependencies([control_flow_ops.Assert(l != cardinality.INFINITE, [msg])]):
  l = tf.identity(l)
```

That way, we avoid the extra tf.cond altogether.

Reply:
Thanks @mdanatg. Updated.